packages/git/src/browser/git-base-widget.ts

```diff
@@ -5,18 +5,68 @@
  * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
  */
 
-import { VirtualWidget } from "@theia/core/lib/browser";
+import { VirtualWidget, SELECTED_CLASS } from "@theia/core/lib/browser";
 import { GitFileStatus, Repository, GitFileChange } from '../common';
 import URI from "@theia/core/lib/common/uri";
 import { GitRepositoryProvider } from "./git-repository-provider";
 import { LabelProvider } from "@theia/core/lib/browser/label-provider";
+import { Message } from "@phosphor/messaging";
+import { Key } from "@theia/core/lib/browser/keys";
+import { SelectionService } from "@theia/core";
+import { ElementExt } from "@phosphor/domutils";
 
-export class GitBaseWidget extends VirtualWidget {
+export enum SelectDirection {
+    NEXT, PREVIOUS
+}
+
+export interface GitFileChangeNode extends GitFileChange {
+    readonly icon: string;
+    readonly label: string;
+    readonly description: string;
+    readonly caption?: string;
+    readonly extraIconClassName?: string;
+    readonly commitSha?: string;
+    selected?: boolean;
+}
+
+export namespace GitFileChangeNode {
+    export function is(node: any): node is GitFileChangeNode {
+        return 'uri' in node && 'status' in node && 'description' in node && 'label' in node && 'icon' in node;
+    }
+}
+
+export abstract class GitBaseWidget<T extends { selected?: boolean }> extends VirtualWidget {
+
+    protected gitNodes: T[];
+    protected abstract readonly scrollContainer: string;
 
     constructor(
         protected readonly repositoryProvider: GitRepositoryProvider,
-        protected readonly labelProvider: LabelProvider) {
+        protected readonly labelProvider: LabelProvider,
+        protected readonly selectionService: SelectionService) {
         super();
+
+        selectionService.onSelectionChanged((c: T) => {
+            c.selected = true;
```

Comment:
could `c` be `undefined`?

Reply:
please can you check, this is the `SelectionService ` of the application, which means that any module may do `selectionService.selection = undefined` and you need to handle it.

Reply:
we should improve types for `SelectionService`, i.e. `Object | undefined` instead of `any`